ast/expressions.go

```diff
@@ -47,21 +51,58 @@ var (
 // ValueExpr is the simple value expression.
 type ValueExpr struct {
 	exprNode
-	// Val is the literal value.
-	Val interface{}
+}
+
+// NewValueExpr creates a ValueExpr with value, and sets default field type.
+func NewValueExpr(value interface{}) *ValueExpr {
+	ve := &ValueExpr{}
+	ve.Data = types.RawData(value)
+	// TODO: make it more precise.
+	switch value.(type) {
+	case nil:
+		ve.Type = types.NewFieldType(mysql.TypeNull)
+	case bool, int64:
+		ve.Type = types.NewFieldType(mysql.TypeLonglong)
+	case uint64:
+		ve.Type = types.NewFieldType(mysql.TypeLonglong)
+		ve.Type.Flag |= mysql.UnsignedFlag
+	case string, UnquoteString:
+		ve.Type = types.NewFieldType(mysql.TypeVarchar)
+		ve.Type.Charset = mysql.DefaultCharset
```

Comment:
Use charset/collation specified in character_set_connection and collation_connection system variables.